lib/handler/cache-handler.js

```diff
@@ -0,0 +1,353 @@
+      // Don't have the headers we need, can't cache
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const maxEntrySize = this.#getMaxEntrySize()
+    const contentLength = Number(headers['content-length'])
+    const currentSize =
+      this.#getSizeOfBuffers(rawHeaders) + (statusMessage?.length ?? 0) + 64
+    if (
+      !Number.isInteger(contentLength) ||
+      contentLength > maxEntrySize ||
+      currentSize > maxEntrySize
+    ) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+    if (!canCacheResponse(statusCode, headers, cacheControlDirectives)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#value = {
+        complete: false,
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        body: [],
+        vary: varyDirectives,
+        size: currentSize,
+        cachedAt: now,
+        staleAt: now + staleAt,
+        deleteAt: now + deleteAt
+      }
+    }
+
+    return this.#handler.onHeaders(
+      statusCode,
+      rawHeaders,
+      resume,
+      statusMessage,
+      headers
+    )
+  }
+
+  /**
+   * @see {DispatchHandlers.onData}
+   *
+   * @param {Buffer} chunk
+   * @returns {boolean}
+   */
+  onData (chunk) {
+    if (this.#value) {
+      this.#value.size += chunk.length
+
+      if (this.#value.size > this.#getMaxEntrySize()) {
+        this.#value = null
+      } else {
+        this.#value.body.push(chunk)
+      }
+    }
+
+    return this.#handler.onData(chunk)
+  }
+
+  /**
+   * @see {DispatchHandlers.onComplete}
+   *
+   * @param {string[] | null} rawTrailers
+   */
+  onComplete (rawTrailers) {
+    if (this.#value) {
+      this.#value.complete = true
+      this.#value.rawTrailers = rawTrailers
+      this.#value.size += this.#getSizeOfBuffers(rawTrailers)
+
+      // If we're still under the max entry size, let's add it to the cache
+      if (this.#getMaxEntrySize() > this.#value.size) {
+        const result = this.#opts.store.put(this.#req, this.#value)
+        if (result && result.constructor.name === 'Promise') {
+          result.catch(err => {
+            throw err
```

Comment:
Currently it will cause `unhandledRejection` as there's not stack to cover it and we will be returning the control to the caller by calling `this.#handler.onComplete` regardless of the promise fulfilling or not.

```suggestion
            handler.onError(err)
```